test/pytest/test_binary_cnn.py

```diff
@@ -12,9 +12,22 @@
 test_root_path = Path(__file__).parent
 
 
-@pytest.mark.parametrize('backend', ['Vivado', 'Vitis', 'Quartus'])
-@pytest.mark.parametrize('io_type', ['io_parallel', 'io_stream'])
-def test_model2(backend, io_type):
+@pytest.mark.parametrize(
+    'backend,io_type,strategy',
+    [
+        ('Quartus', 'io_parallel', 'resource'),
+        ('Quartus', 'io_stream', 'resource'),
+        ('Vivado', 'io_parallel', 'resource'),
+        ('Vivado', 'io_parallel', 'latency'),
+        ('Vivado', 'io_stream', 'latency'),
+        ('Vivado', 'io_stream', 'resource'),
+        ('Vitis', 'io_parallel', 'resource'),
+        ('Vitis', 'io_parallel', 'latency'),
+        ('Vitis', 'io_stream', 'latency'),
+        ('Vitis', 'io_stream', 'resource'),
+    ],
+)
+def test_model2(backend, io_type, strategy):
     x_in = Input(shape=(28, 28, 1))
 
     x = QConv2D(4, (3, 3), kernel_quantizer="binary", name="conv2d_1", kernel_regularizer=l2(0.0001), use_bias=False)(x_in)
```

Comment:
Can bias be enabled? The emphasis here I guess is on the multiplications and pooling precision, but just to confirm it works.

Reply:
I am not sure if bias is supported in binary models, especially if both the weight and input are xnor. I wasn't sure how the bias would work. Let me look at it a bit more. Xnor bias (add or subtract one) didn't seem to make much sense to me, but I am no expert in binary networks. But maybe in the case when only the weights are binary but the inputs aren't it makes sense to have a quantized bias?

Reply:
So 'binary' bias implemented as `XnorPrecisonType` does not work, but I don't think that's really useful. Now `quantized_bits` or standard unquantized bias I think works with fixed or integer inputs and binary (xnor) weights, but I am not sure with xnor inputs and outputs. The complication, if I understand the code, is that the bias addition in the code happens before the cast:
```c++
template <class data_T, class res_T, typename CONFIG_T>
inline typename std::enable_if<std::is_same<data_T, ap_uint<1>>::value &&
                                   std::is_same<typename CONFIG_T::weight_t, ap_uint<1>>::value,
                               ap_int<nnet::ceillog2(CONFIG_T::n_in) + 2>>::type
cast(typename CONFIG_T::accum_t x) {
    return (ap_int<nnet::ceillog2(CONFIG_T::n_in) + 2>)(x - CONFIG_T::n_in / 2) * 2;
}
```
while logically it should happen after. I think the bias is doubled, and the type calculation doesn't include the bias. But let me know if I am wrong. It's easy to lose track when following the code.

So I think I will add bias in all cases except when both the inputs and weights are xnor type.